src/plugins/links/public/components/editor/links_editor.tsx

```diff
@@ -0,0 +1,314 @@
+  {
+    id: LINKS_VERTICAL_LAYOUT,
+    label: LinksLayoutInfo[LINKS_VERTICAL_LAYOUT].displayName,
+  },
+  {
+    id: LINKS_HORIZONTAL_LAYOUT,
+    label: LinksLayoutInfo[LINKS_HORIZONTAL_LAYOUT].displayName,
+  },
+];
+
+const LinksEditor = ({
+  onSaveToLibrary,
+  onAddToDashboard,
+  onClose,
+  initialLinks,
+  initialLayout,
+  parentDashboard,
+  isByReference,
+}: {
+  onSaveToLibrary: (newLinks: Link[], newLayout: LinksLayoutType) => Promise<void>;
+  onAddToDashboard: (newLinks: Link[], newLayout: LinksLayoutType) => void;
+  onClose: () => void;
+  initialLinks?: Link[];
+  initialLayout?: LinksLayoutType;
+  parentDashboard?: DashboardContainer;
+  isByReference: boolean;
+}) => {
+  const toasts = coreServices.notifications.toasts;
+  const editLinkFlyoutRef: React.RefObject<HTMLDivElement> = useMemo(() => React.createRef(), []);
+
+  const [currentLayout, setCurrentLayout] = useState<LinksLayoutType>(
+    initialLayout ?? LINKS_VERTICAL_LAYOUT
+  );
+  const [isSaving, setIsSaving] = useState(false);
+  const [orderedLinks, setOrderedLinks] = useState<Link[]>([]);
+  const [saveByReference, setSaveByReference] = useState(!initialLinks ? true : isByReference);
+
+  const isEditingExisting = initialLinks || isByReference;
+
+  useEffect(() => {
+    if (!initialLinks) {
+      setOrderedLinks([]);
+      return;
+    }
+    setOrderedLinks(memoizedGetOrderedLinkList(initialLinks));
+  }, [initialLinks]);
+
+  const onDragEnd = useCallback(
+    ({ source, destination }) => {
+      if (source && destination) {
+        const newList = euiDragDropReorder(orderedLinks, source.index, destination.index).map(
+          (link, i) => {
+            return { ...link, order: i };
+          }
+        );
+        setOrderedLinks(newList);
+      }
+    },
+    [orderedLinks]
+  );
+
+  const addOrEditLink = useCallback(
+    async (linkToEdit?: Link) => {
+      const newLink = await openLinkEditorFlyout({
+        parentDashboard,
+        link: linkToEdit,
+        ref: editLinkFlyoutRef,
+      });
+      if (newLink) {
+        if (linkToEdit) {
+          setOrderedLinks(
+            orderedLinks.map((link) => {
+              if (link.id === linkToEdit.id) {
+                return { ...newLink, order: linkToEdit.order } as Link;
+              }
+              return link;
+            })
+          );
+        } else {
+          setOrderedLinks([...orderedLinks, { ...newLink, order: orderedLinks.length } as Link]);
+        }
+      }
+    },
+    [editLinkFlyoutRef, orderedLinks, parentDashboard]
+  );
+
+  const hasZeroLinks = useMemo(() => {
+    return orderedLinks.length === 0;
+  }, [orderedLinks]);
+
+  const deleteLink = useCallback(
+    (linkId: string) => {
+      setOrderedLinks(
+        orderedLinks.filter((link) => {
+          return link.id !== linkId;
+        })
+      );
+    },
+    [orderedLinks]
+  );
+
+  return (
+    <>
+      <div ref={editLinkFlyoutRef} />
+      <EuiFlyoutHeader hasBorder>
```

Comment:
For some reason the "add link" flyout flickers for me when I open it, I haven't noticed it for other flyouts


https://github.com/elastic/kibana/assets/7784120/4f151693-e6e5-4fc8-b4d4-96ea2604548a

Reply:
Great catch - something must have changed, because that definitely didn't happen originally. Looking into this!

Reply:
Oof, that was a tough one! Basically, there is a known issue in EUI where tooltips + flyouts **do not** play nice together. This, combined with how we were setting the width of the flyout, was causing the flyout to almost always have this "double open" animation. I was able to fix this by: 
1. Because we are async loading everything (which is something other flyouts aren't necessarily doing), we were actually opening our flyout **too quickly**; so, I added a `50ms` delay **only** when creating a new links panel, since this is the only place where there can be a tooltip on screen **as** the flyout tries to open
2. I set the width of the flyout via the `maxWidth` prop rather than manually setting it in our CSS

These changes where made here: 1aa4fc3db993a78765f09028c4a006858f5166c2